### DISCUSSION TITLE
Using an external crystal oscillator. How to configure?

### DISCUSSION THREAD

**Discussion by mgrosvenor** [SpenceKonde/megaTinyCore#1133] [Q&A]
I have an application which requires better time keeping than the internal clock offers. I'm trying to retrofit a 32.768Khz oscillator on to my board. So far I've had no luck. I've tried various capacitance values between  0pf and 15pf.  I have one doubt though, looking at the data sheet I can see that there are two external clocking options
• 32.768 kHz external crystal oscillator (on `TOSC1` / `TOSC2`) 
• External clock input (on `EXTCLK` aka `PA3`)

When selecting "external" from the list of clocking, is it true the it's the first of these options which is configured?

**Comment by hmeijdam** [comment]
I don't see an option to select the KHz clock as a source in the "clock" menu. Only the MHz sources either internal or external.

If I were in your situation I'd have the system clock just running from the internal oscillator and add a 32.768 KHz crystal to be the clock source for the RTC. Then do the precise timekeeping with the RTC or the PIT.
![image](https://github.com/user-attachments/assets/889aab85-2533-4469-b2aa-60429d9b751c)


[edit] and to your question of configuring TOSC1 as clock source for the RTC, that's done via setting the CLKSEL register.
![image](https://github.com/user-attachments/assets/61020bea-3017-4cf1-8c52-c74cae46a07a)

**Reply by mgrosvenor** [reply]
Thanks for the replies, I’m sorry if the question wasn’t clear. The end state I’m looking for is a 10MHz clock speed for the whole chip, which I’m guessing is synthesised from the external crystal? Or is that not the case?

**Reply by hmeijdam** [reply]
No. the system clock runs from a separate internal oscillator on 16Mhz or 20Mhz that can be further prescaled down. Or you will have to connect an external clock signal. E.g. generated by an active 20Mhz crystal oscillator. You cannot use a passive crystal for the main clock.

You will have to be a lot more precise in describing your actual application and problem, to describe what you actually want.

**Reply by mgrosvenor** [reply]
Thank you for your insights. I was looking at this diagram, which shows that the system clock can be run from the external oscillator: 

![image](https://github.com/user-attachments/assets/dbf52b3b-780a-4dc5-a66c-1f8e3c556e58)

However, I had wrongly assumed that the prescaler was a full cock management unit, which includes a PLL to upscale the clock. It turns out that it's a clock divider only which is disappointing and explains what I'm seeing.

**Comment by hmeijdam** [comment]
You also don't want to forget to actually enable the external 32Khz crystal by setting the "enable" bit in the XOSC32KCTRLA register.

I made an example sketch for the RTC and PIT a while ago and just reworked it to work with an external 32KHz crystal

It wil blink leds on PB0 (via RTC interrupt) and PC0 (via PIT interrupt)

```
void setup() {
  //configure output pins
  VPORTC_DIR |= PIN0_bm;   // set pin PC0 as output
  VPORTB_DIR |= PIN0_bm;   // set pin PB0 as output

  //Configure clock source used by both PIT and RTC
  //RTC_CLKSEL = RTC_CLKSEL_INT32K_gc; /* Internal 32kHz OSC */
  //RTC_CLKSEL = RTC_CLKSEL_INT1K_gc; /* Internal 1kHz OSC */
  RTC_CLKSEL = RTC_CLKSEL_TOSC32K_gc; // When using TOSC32K, don't forget to turn it on!
  //RTC_CLKSEL = RTC_CLKSEL_EXTCLK_gc;  // Use EXTCLK pin as clock source (connect to CLKOUT)

  CCP = CCP_IOREG_gc;    // enable write to protected IO registers (timed sequence)
  CLKCTRL_XOSC32KCTRLA |= CLKCTRL_ENABLE_bm;// turn on External 32Khz OSC (needed for TOSC32K)

  //Configure RTC
  while (RTC_STATUS & RTC_PERBUSY_bm) {} // wait until RTC_PER is available for writing
  RTC_PER = 10; // write period (16bit) to period
  //  RTC_PER = 50000; // write period (16bit) to period
  RTC_INTCTRL |= RTC_OVF_bm; // activate overflow interrupts
  RTC_CTRLA |= RTC_PRESCALER_DIV8192_gc | RTC_RTCEN_bm; // turn on RTC ansd set RTC prescaler factor
  // possible RTC prescaler factors are DIV1, DIV2, DIV4, DIV8, DIV16, ........DIV16384, DIV32768

  //Configure PIT
  RTC_PITINTCTRL |= RTC_PI_bm;    // activate PIT Interrupt enable
  RTC_PITCTRLA   |= RTC_PERIOD_CYC32768_gc | RTC_PITEN_bm;// set PIT period and enable
}
// 32.768KHz / CYCprescale = periodtime in mS
//
//  BIN         NAME        TIME
// 0b00001001  CYC4        0,122 mS
// 0b00011001  CYC8        0,244 mS
// 0b00011001  CYC16       0,488 mS
// 0b00100001  CYC32       0,977 mS
// 0b00101001  CYC64       1,953 mS
// 0b00110001  CYC128      3,906 mS
// 0b00111001  CYC256      7,813 mS
// 0b01000001  CYC512      15,63 mS
// 0b01001001  CYC1024     31,25 mS
// 0b01010001  CYC2048     62,50 mS
// 0b01011001  CYC4096    125,00 mS
// 0b01100001  CYC8192    250,00 mS
// 0b01101001  CYC16384   500,00 mS
// 0b01110001  CYC32768  1000,00 mS


void loop() {} // nothing to do here


ISR(RTC_PIT_vect) {             // RTC Interrupt
  RTC_PITINTFLAGS = RTC_PI_bm;  // clear  Interrupt Flag
  PORTC_OUTTGL = PIN0_bm;       // toggle Pin/C0
}

ISR(RTC_CNT_vect) {             // RTC Interrupt
  RTC_INTFLAGS = RTC_OVF_bm;  // clear  Interrupt Flag
  PORTB_OUTTGL = PIN0_bm;       // toggle Pin/B0
}
```
